rpcs3/Emu/RSX/Program/ProgramStateCache.cpp

```diff
@@ -26,25 +26,26 @@ using namespace program_hash_util;
 
 usz vertex_program_utils::get_vertex_program_ucode_hash(const RSXVertexProgram &program)
 {
-	// 64-bit Fowler/Noll/Vo FNV-1a hash code
-	usz hash = 0xCBF29CE484222325ULL;
+	// Checksum as hash with rotated data
 	const void* instbuffer = program.data.data();
-	usz instIndex = 0;
+	u32 instIndex = 0;
+	usz acc0 = 0;
+	usz acc1 = 0;
 
-	for (unsigned i = 0; i < program.data.size() / 4; i++)
+	do
 	{
-		if (program.instruction_mask[i])
```

Comment:
Maybe this `if` be replaced with a mask on the temporary result of the instruction?

Reply:
It's a good idea, I tested with this loop:

```
	do
	{
		const uint64_t mask = -static_cast<uint64_t>(program.instruction_mask[instIndex]);

		const auto inst = v128::loadu(instbuffer, instIndex);
		usz tmp0 = std::rotr(inst._u64[0], instIndex * 2) & mask;
		acc0 += tmp0;
		usz tmp1 = std::rotr(inst._u64[1], (instIndex * 2) + 1) & mask;
		acc1 += tmp1;

		instIndex++;
	} while (instIndex < (program.data.size() / 4));
	
```
Unfortunately, it's a little slower, before it used around 0.56% in the profiler, now it is using around 0.63% in the profiler. The number of mispredicted branches is too low for the branchless algorithm to be faster. Would likely need to be simdified to be faster.